packages/studio/src/store/slices/pages/createPageSlice.ts

```diff
@@ -160,6 +160,82 @@ export const createPageSlice: SliceCreator<PageSlice> = (set, get) => {
         (component) => component.uuid === activeComponentUUID
       );
     },
+    addSelectedComponentUUID: (selectedUUID: string) => {
+      set((store) => {
+        store.selectedComponentUUIDs.add(selectedUUID);
+      });
+    },
+    addSelectedComponentRect: (
+      selectedUUID: string,
+      rect: DOMRectProperties
+    ) => {
+      set((store) => {
+        store.selectedComponentRectsMap.set(selectedUUID, rect);
+      });
+    },
+    clearSelectedComponents: () => {
+      set((store) => {
+        store.selectedComponentUUIDs = new Set();
+        store.selectedComponentRectsMap = new Map();
+      });
+    },
+    /**
+     * Adds all of the components and their children ordered between the selected component and
+     * the active component (inclusive) to selectedComponentUUIDs.
+     */
+    addShiftSelectedComponentUUIDs: (selectedComponent: ComponentState) => {
+      const selectedUUID = selectedComponent.uuid;
+      const {
+        activeComponentUUID,
+        getActivePageState,
+        addSelectedComponentUUID,
+      } = get();
+      const activePageState = getActivePageState();
+      const componentTree = activePageState?.componentTree;
+      if (selectedUUID === activeComponentUUID) return;
+      if (!componentTree) {
+        throw new Error(
+          `Error adding components: component tree is undefined.`
+        );
+      }
+      if (!activeComponentUUID) {
+        throw new Error(
+          `Error adding components: active component is undefined.`
+        );
+      }
+      const targetComponentUUIDs = [selectedUUID, activeComponentUUID] as [
+        string,
+        string
+      ];
+      const lowestParentUUID = ComponentTreeHelpers.getLowestParentUUID(
+        ...targetComponentUUIDs,
+        componentTree
+      );
+
+      let selected = false;
+      ComponentTreeHelpers.mapComponentTreeParentsFirst(componentTree, (c) => {
+        if (c.parentUUID !== lowestParentUUID) return;
+        const descendants = ComponentTreeHelpers.getDescendants(
+          c,
+          componentTree
+        );
+        const targetIsParent = targetComponentUUIDs.includes(c.uuid);
+        const targetInDescendants = descendants.some((d) =>
+          targetComponentUUIDs.includes(d.uuid)
+        );
+        if (targetIsParent || targetInDescendants) {
+          if (selected || (targetIsParent && targetInDescendants)) {
+            descendants.forEach((d) => addSelectedComponentUUID(d.uuid));
+            addSelectedComponentUUID(c.uuid);
+            selected = false;
+          } else selected = true;
+        }
+        if (selected) {
+          descendants.forEach((d) => addSelectedComponentUUID(d.uuid));
+          addSelectedComponentUUID(c.uuid);
```

Comment:
The inside of this if statement looks identical to the if statement above (if you remove the outside `if`).  I'm wondering if their logic be consolidated any?

Reply:
the shift select function has to select an inclusive range so I'm not sure if it can be consolidated into one. but the similar code can be combined into a function and called twice if that would be better